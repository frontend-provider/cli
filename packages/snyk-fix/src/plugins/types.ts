import { EntityToFix } from '../types';

export type FixHandler = (entities: EntityToFix[]) => Promise<FixHandlerResult[]>;

export interface FixHandlerResult {
  succeeded: EntityToFix[];
  failed: EntityToFix[];
}
