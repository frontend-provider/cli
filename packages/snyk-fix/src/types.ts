import { DepGraphData } from '@snyk/dep-graph';
import { CustomError } from './lib/errors/custom-error';

/* Scan Result
 * this data is returned by the CLI plugins to identify
 * what should be scanned for issues
 */
export interface GitTarget {
  remoteUrl: string;
  branch: string;
}
export interface ContainerTarget {
  image: string;
}

export interface ScanResult {
  readonly identity: Identity;
  readonly facts: Facts[];
  readonly name?: string;
  readonly policy?: string;
  readonly target?: GitTarget | ContainerTarget;
}

export interface Identity {
  type: string;
  targetFile?: string;
  // options used to scan should be here
  args?: { [key: string]: string };
}

export interface Facts {
  type: string;
  data: any;
}

/* Test Result
 * this data is returned on a `snyk test` for supported project types
 * after the relevant plugin extracts dependencies
 */
export interface TestResult {
  readonly issues: Issue[];
  readonly issuesData: IssuesData;
  readonly depGraphData: DepGraphData;
  readonly remediation?: RemediationChanges; // TODO: not yet in the CLI TestResults type
}

export interface Issue {
  pkgName: string;
  pkgVersion?: string;
  issueId: string;
  fixInfo: FixInfo;
}

interface UpgradePath {
  path: UpgradePathItem[];
}

export interface FixInfo {
  upgradePaths?: UpgradePath[];
  isPatchable?: boolean;
  nearestFixedInVersion?: string;
}

interface UpgradePathItem {
  name: string;
  version: string;
  newVersion?: string;
  isDropped?: boolean;
}
export interface IssuesData {
  [issueId: string]: {
    id: string;
    severity: SEVERITY;
    title: string;
  };
}

/* Remediation Data
 * this data is returned on a `snyk test` for supported project types
 */
export interface Upgrade {
  upgradeTo: string; // name@version
}

export interface UpgradeVulns extends Upgrade {
  vulns: string[];
}
export interface UpgradeRemediation extends UpgradeVulns {
  upgrades: string[];
}

export interface PatchRemediation {
  paths: PatchObject[];
}

export interface DependencyUpdates {
  [from: string]: UpgradeRemediation;
}

export interface PinRemediation extends UpgradeVulns {
  isTransitive: boolean;
}

export interface DependencyPins {
  [name: string]: PinRemediation;
}

// Remediation changes to be applied to the project,
// including information on all and unresolved issues.
export interface RemediationChanges {
  unresolved: IssueData[];
  upgrade: DependencyUpdates;
  patch: {
    [name: string]: PatchRemediation;
  };
  ignore: unknown;
  pin: DependencyPins;
}

export interface IssueData {
  id: string;
  packageName: string;
  version: string;
  moduleName?: string;
  below: string; // Vulnerable below version
  semver: {
    vulnerable: string | string[];
    vulnerableHashes?: string[];
    vulnerableByDistro?: {
      [distroNameAndVersion: string]: string[];
    };
  };
  patches: Patch[];
  isNew: boolean;
  description: string;
  title: string;
  severity: SEVERITY;
  fixedIn: string[];
  legalInstructions?: string;
  reachability?: REACHABILITY;
}

interface Patch {
  version: string;
  id: string;
  urls: string[];
  modificationTime: string;
}

export enum REACHABILITY {
  FUNCTION = 'function',
  PACKAGE = 'package',
  NOT_REACHABLE = 'not-reachable',
  NO_INFO = 'no-info',
}

export interface PatchObject {
  [name: string]: {
    patched: string;
  };
}

export enum SEVERITY {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

/* End Remediation Data
 */

/* Snyk fix types
 * Types for concepts introduced as part of this lib
 */

export type SupportedScanTypes = 'pip';

export interface Workspace {
  path: string;
  readFile: (path: string) => Promise<string>;
  writeFile: (path: string, content: string) => Promise<void>;
}
export interface EntityToFix {
  readonly workspace: Workspace;
  readonly scanResult: ScanResult;
  readonly testResult: TestResult;
  readonly options: CliTestOptions;
}

// Partial CLI test options interface
// defining only what is used by @snyk/fix
// add more as needed
export interface PythonTestOptions {
  command?: string; // python interpreter to use for python tests
  dev?: boolean;
}
export type CliTestOptions = PythonTestOptions;
export interface WithError<Original> {
  original: Original;
  error: CustomError;
  tip?: string;
}

export interface WithFixChangesApplied<Original> {
  original: Original;
  changes: FixChangesSummary[];
}

export interface WithUserMessage<Original> {
  original: Original;
  userMessage: string;
}

export type FixChangesSummary = FixChangesSuccess | FixChangesError;

interface FixChangesSuccess {
  success: true;
  userMessage: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

interface FixChangesError {
  success: false;
  userMessage: string;
  reason: string;
  tip?: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

export interface ErrorsByEcoSystem {
  [ecosystem: string]: { originals: EntityToFix[]; userMessage: string };
}
export interface FixOptions {
  dryRun?: boolean;
  quiet?: boolean;
  stripAnsi?: boolean;
}

export interface FixedMeta {
  fixed: number;
  failed: number;
  fixableIssues: number;
  fixedIssues: number;
  totalIssues: number;
}
