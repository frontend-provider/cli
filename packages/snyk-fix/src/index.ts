import * as debugLib from 'debug';
import { loadPlugin } from './plugins/load-plugin';
import { FixHandlerResult } from './plugins/types';

import { EntityToFix } from './types';

const debug = debugLib('snyk-fix:main');

export async function fix(
  entities: EntityToFix[],
): Promise<FixHandlerResult[]> {
  debug(`Requested to fix ${entities.length} projects.`);
  const entitiesPerType = groupEntitiesPerType(entities);
  const allResults: FixHandlerResult[] = [];
  // TODO: pMap this?
  for (const type of Object.keys(entitiesPerType)) {
    const handler = loadPlugin(type);
    const results = await handler(entitiesPerType[type]);
    allResults.push(results);
  }

  return allResults;
}

function groupEntitiesPerType(
  entities: EntityToFix[],
): {
  [type: string]: EntityToFix[];
} {
  const entitiesPerType: {
    [type: string]: EntityToFix[];
  } = {};
  for (const entity of entities) {
    const type = entity.scanResult.identity.type;
    if (entitiesPerType[type]) {
      entitiesPerType[type].push(entity);
      continue;
    }
    entitiesPerType[type] = [entity];
  }
  return entitiesPerType;
}
