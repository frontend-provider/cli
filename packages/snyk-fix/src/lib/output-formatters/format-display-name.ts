import * as pathLib from 'path';
import { Identity } from '../../types';

export function formatDisplayName(
  path: string,
  identity: Pick<Identity, 'type' | 'targetFile'>,
): string {
  if (!identity.targetFile) {
    return `${identity.type} project`;
  }
  // show paths relative to where `snyk fix` is running
  return pathLib.relative(
    process.cwd(),
    pathLib.join(path, identity.targetFile),
  );
}
