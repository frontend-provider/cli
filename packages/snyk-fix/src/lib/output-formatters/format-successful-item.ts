import * as chalk from 'chalk';

import { EntityToFix, FixChangesSummary } from '../../types';
import { PADDING_SPACE } from './show-results-summary';

/*
 * Generate formatted output that describes what changes were applied, which failed.
 */
export function formatChangesSummary(
  entity: EntityToFix,
  changes: FixChangesSummary[],
): string {
  return `${PADDING_SPACE}${
    entity.scanResult.identity.targetFile
  }\n${changes.map((c) => formatAppliedChange(c)).join('\n')}`;
}

// TODO:
// write test for these
function formatAppliedChange(change: FixChangesSummary): string | null {
  if (change.success === true) {
    return `${PADDING_SPACE}${chalk.green('âœ”')} ${change.userMessage}`;
  }
  if (change.success === false) {
    return `${PADDING_SPACE}${chalk.red('x')} ${chalk.red(
      change.userMessage,
    )}\n${PADDING_SPACE}Reason:${PADDING_SPACE}${change.reason}${
      change.tip ? `.\n${PADDING_SPACE}Tip:     ${change.tip}` : undefined
    }`;
  }
  return '';
}
