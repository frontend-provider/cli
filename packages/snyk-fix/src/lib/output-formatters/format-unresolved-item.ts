import * as chalk from 'chalk';

import { EntityToFix } from '../../types';
import { formatDisplayName } from './format-display-name';
import { PADDING_SPACE } from './show-results-summary';

export function formatUnresolved(
  entity: EntityToFix,
  userMessage: string,
  tip?: string,
): string {
  const name = formatDisplayName(
    entity.workspace.path,
    entity.scanResult.identity,
  );
  const tipMessage = tip ? `\n${PADDING_SPACE}Tip:     ${tip}` : '';
  const errorMessage = `${PADDING_SPACE}${name}\n${PADDING_SPACE}${chalk.red(
    'âœ–',
  )} ${chalk.red(userMessage)}`;
  return errorMessage + tipMessage;
}
