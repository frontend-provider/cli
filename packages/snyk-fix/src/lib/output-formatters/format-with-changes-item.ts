import * as chalk from 'chalk';

import { EntityToFix, FixChangesSummary } from '../../types';
import { formatDisplayName } from './format-display-name';
import { PADDING_SPACE } from './show-results-summary';

/*
 * Generate formatted output that describes what changes were applied, which failed.
 */
export function formatChangesSummary(
  entity: EntityToFix,
  changes: FixChangesSummary[],
): string {
  return `${PADDING_SPACE}${formatDisplayName(
    entity.workspace.path,
    entity.scanResult.identity,
  )}\n${changes.map((c) => formatAppliedChange(c)).join('\n')}`;
}

function formatAppliedChange(change: FixChangesSummary): string | null {
  if (change.success === true) {
    return `${PADDING_SPACE}${chalk.green('âœ”')} ${change.userMessage}`;
  }
  if (change.success === false) {
    return `${PADDING_SPACE}${chalk.red('x')} ${chalk.red(
      change.userMessage,
    )}\n${PADDING_SPACE}Reason:${PADDING_SPACE}${change.reason}${
      change.tip ? `.\n${PADDING_SPACE}Tip:     ${change.tip}` : undefined
    }`;
  }
  return '';
}
