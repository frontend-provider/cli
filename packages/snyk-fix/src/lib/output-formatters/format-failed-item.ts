import { FailedToFix, isWithError } from '../../plugins/types';

import { convertErrorToUserMessage } from '../errors/error-to-user-message';
import { formatChangesSummary } from './format-with-changes-item';
import { formatUnresolved } from './format-unresolved-item';

export function formatFailed(failed: FailedToFix): string {
  if (isWithError(failed)) {
    return formatUnresolved(
      failed.original,
      convertErrorToUserMessage(failed.error),
      failed.tip,
    );
  }
  return formatChangesSummary(failed.original, failed.changes);
}
