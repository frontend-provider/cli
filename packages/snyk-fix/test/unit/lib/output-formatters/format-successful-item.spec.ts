import stripAnsi = require('strip-ansi');
import { formatChangesSummary } from '../../../../src/lib/output-formatters/format-successful-item';
import { FixChangesSummary } from '../../../../src/types';
import { generateEntityToFix } from '../../../helpers/generate-entity-to-fix';

describe('format successful item', () => {
  it('successful item & changes formatted', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const changesSummary: FixChangesSummary[] = [
      {
        success: true,
        userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
        issueIds: ['vuln-2'],
      },
    ];
    const res = await formatChangesSummary(entity, changesSummary);
    expect(stripAnsi(res)).toMatchSnapshot();
  });
});
