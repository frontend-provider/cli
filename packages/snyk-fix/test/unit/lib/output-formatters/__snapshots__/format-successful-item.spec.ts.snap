// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format successful item successful item & changes formatted 1`] = `
"  requirements.txt
  âœ” Upgraded Django from 1.6.1 to 2.0.1"
`;
